service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+        this.certificateRepository = certificateRepository;
+        this.tagRepository = tagRepository;
+        this.certificateDTOValidator = certificateDTOValidator;
+        this.pageQueryDTOValidator = pageQueryDTOValidator;
+    }
+
+    @Override
+    public CertificateDTO getById(Long id) {
+        Certificate certificate = certificateRepository.getById(id).orElseThrow(() -> new EntityNotFoundException(" (id = " + id + ")"));
+        addTagsToCertificate(certificate);
+        return CertificateConverterDTO.convertToDto(certificate);
+    }
+
+
+    @Override
+    @Transactional
+    public CertificateDTO create(CertificateDTO certificateDTO) {
+        certificateDTOValidator.cleanErrorMessages();
+        if (!certificateDTOValidator.isValid(certificateDTO)) {
+            throw new ValidatorException(certificateDTOValidator.getErrorMessage());
+        }
+        String name = certificateDTO.getName();
+        if (certificateRepository.getByName(name).isPresent()) {
+            throw new EntityAlreadyExistsException(" (name = " + name + ")");
+        }
+        Certificate certificate = CertificateConverterDTO.convertToEntity(certificateDTO);
+        checkTags(certificate);
+        certificate = certificateRepository.save(certificate);
+        tagRepository.createCertificateTags(certificate);
+        certificate = certificateRepository.getById(certificate.getId()).orElseThrow(EntityNotFoundException::new);
+        addTagsToCertificate(certificate);
+        return CertificateConverterDTO.convertToDto(certificate);
+    }
+
+    @Override
+    @Transactional
+    public CertificateDTO update(CertificateDTO certificateDTO) {
+        certificateDTOValidator.cleanErrorMessages();
+        if (!certificateDTOValidator.isValid(certificateDTO)) {
+            throw new ValidatorException(certificateDTOValidator.getErrorMessage());
+        }
+        String name = certificateDTO.getName();
+        if (certificateRepository.getByName(name).isPresent()
+                && !name.equals(certificateRepository.getById(certificateDTO.getId()).get().getName())) {
+            throw new EntityAlreadyExistsException(" (name = " + name + ")");
+        }
+        Certificate certificate = CertificateConverterDTO.convertToEntity(certificateDTO);
+        certificateRepository.deleteCertificateTags(certificate.getId());
+        checkTags(certificate);
+        certificate = certificateRepository.update(certificate);
+        tagRepository.createCertificateTags(certificate);
+        certificate = certificateRepository.getById(certificate.getId()).orElseThrow(EntityNotFoundException::new);
+        addTagsToCertificate(certificate);
+        return CertificateConverterDTO.convertToDto(certificate);
+    }
+
+    private void checkTags(Certificate certificate) {
```

Comment:
This method not only checks tags and also creates new ones invoking `createNewTags`. So its name has to reflect this. E.g. `createNotExistingTags` looks better. Also, you should refactor this method and move logic from `Set<Tag> createNewTags(Set<Tag> tags)` and `Set<Tag> createCertificateTags(Certificate certificate)` to this method or any place in service layer.